compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/HeaderProcessor.kt

```diff
@@ -396,18 +396,34 @@ internal class WithIndexLoopHeader(
             ) {
                 indexVariable = nestedLoopHeader.inductionVariable
                 ownsIndexVariable = false
+                incrementIndexStatement = null
             } else {
                 indexVariable = scope.createTemporaryVariable(
                     irInt(0),
                     nameHint = "index",
                     isMutable = true
                 )
                 ownsIndexVariable = true
+                // `index++` during iteration initialization
+                // TODO: MUSTDO: Check for overflow for Iterable and Sequence (call to checkIndexOverflow()).
+                val plusFun = indexVariable.type.getClass()!!.functions.first {
```

Comment:
What problem with `checkIndexOverflow` support>

Reply:
Looks like `checkIndexOverflow` should be check in another place, otherwise exception would be thrown too early

Reply:
I don't follow, where should the check be? When I looked at the stdlib code, this looks like the right place to do it. It's always `checkIndexOverflow(index++)`

Reply:
The reason I didn't rush to add the check in is because it's also currently broken current JVM backend. https://youtrack.jetbrains.com/issue/KT-34665

It's not hard to add it though.

Reply:
Did you mean that loop optimization is broken in old JVM backend in same case?  Then is should be fixed in 1.4 (maybe with LC ticket cc @zarechenskiy @dnpetrov )
Look like that proper one behaviour should be supported for K/N and K/JS in any way. cc @homuroll 

My point was that such check should be performed after loop body execution for last unoverflow value (e.g. MAX_INT). But if it not presented it doesn't make any sence except maybe mention in comment

Reply:
Yes it's also broken in old JVM backend, that's what KT-34665 is for. I personally don't think it should require LC approval, because the addition of the check was already approved (https://youtrack.jetbrains.com/issue/KT-16097) and the implementation was simply missed for the optimized for-loop case.

If you check the index value (`if (index < 0) throw...`) _before_ incrementing then that would be equivalent to `checkIndexOverflow(index++)`. The implementation of `checkIndexOverflow` checks if the index is negative.